### ISSUE TITLE
Type hints in black (enhancement)

### ISSUE THREAD

**Issue by brianharvey** [jmoenig/Snap--Build-Your-Own-Blocks#1689]
People get confused about whether the type hints in the block editor are part of the variable names, especially about #, which could very plausibly be part of a name such as "color #."  To fix this, I suggest putting the type hint in black:
![hint](https://cloud.githubusercontent.com/assets/3885074/24116096/c9b45cae-0d62-11e7-8ba6-8ddbe60daecf.png)
(Sorry my Photoshop skilz aren't up to the shadowing...)

If making a block of type Variables, so the input blobs are zebra colored and lettered in black, the hint would be in white.

**Comment by jmoenig** [comment]
Sigh. Let's not introduce yet more typography and let's not assign more meaning to colors.

People *give* this name. People can *click* on the thing and see the name. People can drag off a copy and that makes it clear what the name is. If there is confusion there is also ample opportunity to discover more and find out the truth. Even without reading the manual. Stuff like this...

**Comment by brianharvey** [comment]
Sorry, I should have clarified the context better.  I'm a kid.  I haven't _written_ anything; I'm looking at a _picture_ of a script in the curriculum that I'm supposed to copy-and-embellish.

And, yeah, after the fifth or sixth time I probably get it.  But until then it's just one more of the gazillion little details that I don't know anything about, and I have no way of knowing which are critical.  And anything we can do to help would be great, I think.

**Comment by cycomachead** [comment]
> Sigh. Let's not introduce yet more typography and let's not assign more meaning to colors.

I tend to agree with this, especially about color and mostly about typography. 


> People get confused about whether the type hints in the block editor are part of the variable names,

Is also quite correct. In addition to Brian's context, I have seen students confused about where the # comes from. Even though people can discover it's just added by Snap!, I've seen students try to re-edit a variable name and still confused why something is there. This seems to be mostly problematic with the # and the ?, since they are typable characters, whereas the sprite icon or the list icon aren't on the keyboard, so there's less confusion typically.


What if we separate out the type hints a bit more, but in a different method.
Two possible ideas could be:
1. Use a segmented control - that would still be a pill, but it would be "variable | #" where the | is an actual line the entire vertical height. (Kind of like this: http://getbootstrap.com/components/#input-groups-buttons-segmented, only we need to make sure not to make the type look clickable (maybe?))
2. Put the type icon inside some circle / box. I hope this doesn't count as more 'typography', but I do think it would separate things a bit more. 

Separating them a bit isn't such a bad thing, I think. Type annotations are awesome (and now that Python has them I almost want to write more of it. Almost.) Separation, for me, gives you two distinct benefits. 

Firstly, it makes your aware that these two things are independent (i.e. The "#" is not part of the name), and so you notice it. However, because it's separate, in its own you, that also gives you the power to consciously ignore the type hints until you're ready to use them. (which is typically what I tell students when first starting out making blocks, thought not for very long.)

**Comment by jmoenig** [comment]
Yeah, you're both right in pointing out that we need to iterate over the design to make it more clear that the hint is just a *hint* and not part of the name. I like the idea of a graphical (non-character) separator somewhat better than typography. If I remember right the details of changing this mechanism might be a bit hairy, but I'm willing to dive into this once the current flurry of getting 4.1 ready has somewhat subsided.

**Comment by brianharvey** [comment]
You know, the _best_ way to do this would probably be to use a picture of what the input slot will actually look like in the block.   ![foo](https://user-images.githubusercontent.com/3885074/27695379-b54734ae-5cbc-11e7-971d-272a9b97cd2a.png) ![baz](https://user-images.githubusercontent.com/3885074/27695387-bcc3cd96-5cbc-11e7-8482-956e95665002.png)  ![quux](https://user-images.githubusercontent.com/3885074/27695553-3f4b1972-5cbd-11e7-804d-313f71acc80c.png)

**Comment by cycomachead** [comment]
Funny. I’d considered that but rejected the idea. The main problem is that this seems to lead to the confusion that something might be interactive. 

I would be more amenable to this idea if it were a way to specify default inputs, but that’s a larger change. 

There might still be a way to make this work, but it feels like it would add a lot of visual clutter to the block declaration. 
-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by brianharvey** [comment]
Yeah, but it'd mean people wouldn't have to learn _two_ representations for each type.  Maybe if they were ghosted out, like inherited blocks?

**Comment by cycomachead** [comment]
> Or maybe with a colon:

I missed this earlier. I also thought about that. It's fairly common amongst things like Haskell, Swift, Python's annotations and TypeScript. However, it think this leads to the same problems as any other test representation. And it's definitely typography. I, personally, don't think it offers much of a separation. It's notation that you want to read, because it is readable.


> it'd mean people wouldn't have to learn two representations for each type.

This is a good point. I'm not sure how hard they are to learn currently, but we certainly can make  that easier. Exposing the icon in he input type picker would reduce confusion, too. 
A greyed out representation might make sense. I do still think that it's a lot to look at and process though.
